File: Euler#12.highly_divisible_triangular_numbers.rb
@best_div_count = nil 
def tri_num(n)
  return 1 if n.eql?(1)
  return 3 if n.eql?(3)
  nth = (1..n).inject { | sum, num | sum += num } 
  return nth
end 

def div_counter(num)
  @div_count = 0 
  for divisor in 1..Math.sqrt(num)
    @div_count += 1 if ( num % divisor ).eql?(0)
    if @best_div_count.eql?(nil) || @div_count > @best_div_count 
      @best_div_count = @div_count 
      end 
  end
  return @div_count
end 

def step(max)
  @div_count = 0 
  num = 1 
    until @div_count >= max 
    @div_count = (div_counter(tri_num(num)) * 2)
    num += 1 
  end 
puts "Nth: #{num}, Value: #{tri_num(num-1)}"
end 

step(500)

File: Euler#12.highly_divisible_triangular_numbers.js
let bestDivCount = null;

function triNum(n) {
  if (n === 1) return 1;
  if (n === 3) return 3;
  let nth = (n * (n + 1)) / 2; // using the formula for triangular numbers
  return nth;
}

function divCounter(num) {
  let divCount = 0;
  for (let divisor = 1; divisor <= Math.sqrt(num); divisor++) {
    if (num % divisor === 0) {
      divCount++;
      if (num / divisor !== divisor) divCount++; // count the other divisor if it's not a square root
    }
  }
  if (!bestDivCount || divCount > bestDivCount) bestDivCount = divCount;
  return divCount;
}

function findByMaxDivisors(max) {
  let num = 1;
  let divCount = 0;
  while (divCount < max) {
    divCount = divCounter(triNum(num));
    num++;
  }
  console.log(`Nth: ${num - 1}, Value: ${triNum(num - 1)}`);
}

findByMaxDivisors(500);
